### PR TITLE
[ETCM-355] Send fork id to peers

### PR DESCRIPTION
**Pull request by lukasz-golebiewski**
# Description
First part of https://eips.ethereum.org/EIPS/eip-2364. Extends the ETH64 protocol Status by a fork id field and broadcasts it to peers during handshake 
 
Does not yet contain peer fork id validation

### CODE REVIEW

File: src/main/scala/io/iohk/ethereum/network/handshaker/EtcHelloExchangeState.scala [hunk 1]
```diff
@@ -27,13 +27,15 @@ case class EtcHelloExchangeState(handshakerConfiguration: EtcHandshakerConfigura
   override def applyResponseMessage: PartialFunction[Message, HandshakerState[PeerInfo]] = { case hello: Hello =>
     log.debug("Protocol handshake finished with peer ({})", hello)
     // FIXME in principle this should be already negotiated
-    Capability.negotiate(hello.capabilities.toList, handshakerConfiguration.capabilities) match {
+    Capability.negotiate(hello.capabilities.toList, handshakerConfiguration.blockchainConfig.capabilities) match {
       case Some(ProtocolVersions.ETC64) => EtcNodeStatus64ExchangeState(handshakerConfiguration)
       case Some(ProtocolVersions.ETH63) => EtcNodeStatus63ExchangeState(handshakerConfiguration)
```
Comment (dzajkowski) [inline src/main/scala/io/iohk/ethereum/network/handshaker/EtcHelloExchangeState.scala]: request: do you mind changing this to Eth? I missed it and it's confusing this way.

**Comment by dzajkowski** [comment]
overall looks good to me. question: would you be able to deploy to staging to see how it behaves? I guess running it locally against ETC mainnet would also be a good idea.

**Review by leo-bogastry — APPROVED** [review]
I thinks it's looking good

**Review by dzajkowski — APPROVED** [review]

**Comment by lukasz-golebiewski** [comment]
Waiting before merging for https://github.com/input-output-hk/mantis/pull/1029

**Review by AurelienRichez — APPROVED** [review]

File: src/main/scala/io/iohk/ethereum/blockchain/sync/regular/BlockBroadcast.scala [hunk 2]
```diff
@@ -45,10 +45,12 @@ class BlockBroadcast(val etcPeerManager: ActorRef) {
...
   private def broadcastNewBlock(blockToBroadcast: BlockToBroadcast, peers: Map[PeerId, PeerWithInfo]): Unit =
     obtainRandomPeerSubset(peers.values.map(_.peer).toSet).foreach { peer =>
-      val message: MessageSerializable =
-        if (peers(peer.id).peerInfo.remoteStatus.protocolVersion.toByte == ProtocolVersions.ETC64.version)
-          blockToBroadcast.as64
-        else blockToBroadcast.as63
+      val remoteStatus = peers(peer.id).peerInfo.remoteStatus
+
+      val message: MessageSerializable = remoteStatus.protocolFamily match {
+        case ETH => blockToBroadcast.as63
```
Comment (dzajkowski) [inline src/main/scala/io/iohk/ethereum/blockchain/sync/regular/BlockBroadcast.scala:51]: would be so cool to have this fail if a new block type is introduced. just a thought, no great ideas.

File: src/main/scala/io/iohk/ethereum/network/handshaker/EtcHelloExchangeState.scala [hunk 3]
```diff
@@ -31,18 +31,15 @@ case class EtcHelloExchangeState(handshakerConfiguration: EtcHandshakerConfigura
...
-      }
-      case Some(ProtocolVersions.ETH63) => {
+      case Some(ProtocolVersions.ETH63) =>
         log.debug("Negotiated protocol version with client {} is eth/63", hello.clientId)
         EthNodeStatus63ExchangeState(handshakerConfiguration)
-      }
-      case Some(ProtocolVersions.ETH64) => {
+      case Some(ProtocolVersions.ETH64) =>
         log.debug("Negotiated protocol version with client {} is eth/64", hello.clientId)
         EthNodeStatus64ExchangeState(handshakerConfiguration)
```
Comment (dzajkowski) [inline src/main/scala/io/iohk/ethereum/network/handshaker/EtcHelloExchangeState.scala:42]: could you also fix the 'orElse' clause that I introduced (`orElse(WireDecoder)`)

File: src/main/scala/io/iohk/ethereum/blockchain/sync/regular/BlockBroadcast.scala (hunk 2 above)
Comment (lukasz-golebiewski) [reply src/main/scala/io/iohk/ethereum/blockchain/sync/regular/BlockBroadcast.scala:51]: Not sure what you mean here

File: src/main/scala/io/iohk/ethereum/network/handshaker/EtcHelloExchangeState.scala (hunk 3 above)
Comment (lukasz-golebiewski) [reply src/main/scala/io/iohk/ethereum/network/handshaker/EtcHelloExchangeState.scala:42]: I would like to do this in a follow up PR, this one has become quite old and heavy

File: src/main/scala/io/iohk/ethereum/blockchain/sync/regular/BlockBroadcast.scala (hunk 2 above)
Comment (dzajkowski) [reply src/main/scala/io/iohk/ethereum/blockchain/sync/regular/BlockBroadcast.scala:51]: thinking out loud:
```scala
  private def broadcastNewBlock(blockToBroadcast: BlockToBroadcast, peers: Map[PeerId, PeerWithInfo]): Unit =
    obtainRandomPeerSubset(peers.values.map(_.peer).toSet).foreach { peer =>
      val remoteStatus = peers(peer.id).peerInfo.remoteStatus

      val message: MessageSerializable = {
        val capOpt = Capabilities.of(remoteStatus.protocolFamily, remoteStatus.protocolVersion)
        capOpt match {
          case Capabilities.Eth63Capability => blockToBroadcast.as63
          case Capabilities.Eth64Capability => blockToBroadcast.as63
          case Capabilities.Etc64Capability => blockToBroadcast.as64
          case None                         => throw new RuntimeException("Unknown capability, cannot broadcast block")
        }
      }

...

    def of(pf: ProtocolFamily, v: Int): Option[Capability] =
      (pf, v) match {
        case (ProtocolFamily.ETH, 63) => Some(Eth63Capability)
        case (ProtocolFamily.ETH, 64) => Some(Eth64Capability)
        case (ProtocolFamily.ETC, 64) => Some(Etc64Capability)
        case _                        => None
      }
```

and an ADT out of the capabilities?

**Pull request merged**